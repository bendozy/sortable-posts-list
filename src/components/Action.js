import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import timeTravel from '../actions/actions';

export const Action = ({
  action: { title, from, to },
  action,
  timeTravel: timeTravelAction,
  index,
}) => {
  // const [style, setStyle] = useState({
  //   opacity: 0,
  //   transition: 'all 2s ease',
  // });

  // useEffect(() => {
  //   setStyle({
  //     opacity: 1,
  //     transition: 'all 2s ease',
  //   });

  //   return () => {
  //     // setStyle({
  //     //   opacity: 1,
  //     //   transition: 'all 2s ease',
  //     // });
  //     setTimeout(() => {
  //       console.log('removing');
  //     }, 4000);
  //   };
  // }, [action]);

  return (
    <div
      className="Action bg-white rounded p-3 shadow border flex flex-wrap"
      // style={style}
    >
      <div className="sm:my-3 my-1 sm:text-base text-sm w-3/4 text-gray-700">
        {`Moved ${title} from index ${from} to index ${to}`}
      </div>
      <button
        onClick={() => timeTravelAction(index)}
        type="button"
        className="h-12 m-1 bg-green-200 rounded text-sm shadow sm:text-base w-1/5 text-gray-700"
      >
        Time Travel
      </button>
    </div>
  );
};

Action.propTypes = {
  action: PropTypes.shape({
    title: PropTypes.string.isRequired,
    from: PropTypes.number.isRequired,
    to: PropTypes.number.isRequired,
  }).isRequired,
  index: PropTypes.number.isRequired,
  timeTravel: PropTypes.func.isRequired,
};

const mapDispatchToProps = {
  timeTravel,
};

export default connect(null, mapDispatchToProps)(Action);
