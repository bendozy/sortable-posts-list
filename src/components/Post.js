import React from 'react';
import PropTypes from 'prop-types';
import { UP, DOWN } from '../constants';

// React Flip need the child component to not be stateless
/* eslint-disable react/prefer-stateless-function */
export default class Post extends React.Component {
  render() {
    const {
      post: { title },
      index,
      post,
      lastIndex,
      movePost: movePostAction,
    } = this.props;

    return (
      <div className="Post bg-white shadow rounded mb-6 flex justify-between">
        <div className="Post-title py-6 px-2 text-md">{title}</div>
        <div className="Post-buttons flex flex-col pr-2 justify-center">
          {index !== 0 && (
            <button
              type="button"
              className="Post-button moveUp"
              onClick={() => movePostAction({ direction: UP, post, index })}
            >
              <i className="text-2xl fa fa-angle-up" />
            </button>
          )}
          {index !== lastIndex && (
            <button
              type="button"
              className="Post-button moveDown"
              onClick={() => movePostAction({ direction: DOWN, post, index })}
            >
              <i className="text-2xl fa fa-angle-down" />
            </button>
          )}
        </div>
      </div>
    );
  }
}

Post.propTypes = {
  post: PropTypes.shape({
    title: PropTypes.string.isRequired,
  }).isRequired,
  index: PropTypes.number.isRequired,
  lastIndex: PropTypes.number.isRequired,
  movePost: PropTypes.func.isRequired,
};
