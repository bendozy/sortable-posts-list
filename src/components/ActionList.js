import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import FlipMove from 'react-flip-move';
import Action from './Action';
import timeTravelAction from '../actions/actions';

export const ActionList = ({ actions, timeTravel }) => (
  <div className="ActionList">
    <div className="bg-white rounded shadow-md overflow-hidden">
      <div className="ActionList-title font-semibold bg-white p-4 text-gray-800 text-xl">List of actions commited</div>
      <div className="ActionList-items bg-gray-300 p-6 border border-gray-300 overflow-hidden shadow-md">
        <FlipMove
          duration={300}
        >
          {actions.map((action, index) => (
            <Action
              key={action.key}
              action={action}
              index={index}
              timeTravel={timeTravel}
            />
          )).reduce((acc, val) => [val].concat(acc), [])}
          {actions.length === 0 && (
            <div className="ActionList-loading bg-white p-4 rounded shadow-md text-gray-500">
              <span>No actions commited</span>
            </div>
          )}
        </FlipMove>
      </div>
    </div>
  </div>
);

ActionList.propTypes = {
  actions: PropTypes.arrayOf(PropTypes.object).isRequired,
  timeTravel: PropTypes.func.isRequired,
};

export const mapStateToProps = ({ actions: { actions } }) => ({ actions });

export const mapDispatchToProps = {
  timeTravel: timeTravelAction,
};

export default connect(mapStateToProps, mapDispatchToProps)(ActionList);
