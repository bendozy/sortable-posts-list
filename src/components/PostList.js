import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import FlipMove from 'react-flip-move';
import { connect } from 'react-redux';
import Post from './Post';
import { fetchPosts, movePost } from '../actions/posts';

import spinner from '../assets/Spinner.svg';

import '../styles/PostList.css';

export const PostList = ({
  posts: {
    isLoading,
    posts,
  },
  fetchPosts: fetchPostsAction,
  movePost: movePostAction,
}) => {
  useEffect(() => {
    fetchPostsAction();
  }, [fetchPostsAction]);


  const postCards = posts.map((post, index) => (
    <Post
      key={post.id}
      post={post}
      index={index}
      lastIndex={posts.length - 1}
      movePost={movePostAction}
    />
  ));

  return (
    <div className="PostList">
      <h1 className="PostList-title mb-6 font-bold text-3xl text-white">Sortable Post List</h1>
      {isLoading && <img src={spinner} className="PostList-spinner" alt="loading-logo" />}
      {!isLoading && (
        <div className="PostList-items">
          <FlipMove duration={300}>
            {postCards}
          </FlipMove>
        </div>
      )}
    </div>
  );
};

PostList.propTypes = {
  fetchPosts: PropTypes.func.isRequired,
  posts: PropTypes.shape({
    isLoading: PropTypes.bool.isRequired,
    posts: PropTypes.arrayOf(PropTypes.object),
  }).isRequired,
  movePost: PropTypes.func.isRequired,
};

export const mapStateToProps = ({ posts }) => ({ posts });

export const mapDispatchToProps = {
  fetchPosts,
  movePost,
};

export default connect(mapStateToProps, mapDispatchToProps)(PostList);
