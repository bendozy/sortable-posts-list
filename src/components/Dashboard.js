import React from 'react';
import PostList from './PostList';
import ActionList from './ActionList';

const Dashboard = () => (
  <div id="Dashboard">
    <PostList />
    <ActionList />
  </div>
);

export default Dashboard;
