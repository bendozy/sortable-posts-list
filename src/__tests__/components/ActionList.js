import React from 'react';
import toJson from 'enzyme-to-json';
import FlipMove from 'react-flip-move';
import { ActionList, mapStateToProps } from '../../components/ActionList';
import Action from '../../components/Action';

const timeTravel = jest.fn();

describe('<ActionList />', () => {
  let wrapper;

  it('should render a loading component', () => {
    const actions = [];

    wrapper = shallow(
      <ActionList actions={actions} timeTravel={timeTravel} />,
    );

    expect(wrapper.find('.ActionList')).toHaveLength(1);
    expect(wrapper.find('.ActionList-loading')).toHaveLength(1);
    expect(wrapper.find('.ActionList-loading').childAt(0).text()).toBe('No actions commited');
    expect(wrapper.find('.ActionList-title')).toHaveLength(1);
    expect(wrapper.find(FlipMove)).toHaveLength(1);
    expect(wrapper.find(Action)).toHaveLength(0);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render a component with action list items', () => {
    const actions = [
      {
        title: 'test1',
        from: 0,
        to: 1,
        key: 0,
      },
      {
        title: 'test2',
        from: 1,
        to: 2,
        key: 1,
      },
    ];

    wrapper = shallow(
      <ActionList actions={actions} timeTravel={timeTravel} />,
    );

    expect(wrapper.find('.ActionList')).toHaveLength(1);
    expect(wrapper.find('.ActionList-loading')).toHaveLength(0);
    expect(wrapper.find('.ActionList-title')).toHaveLength(1);
    expect(wrapper.find(FlipMove)).toHaveLength(1);
    expect(wrapper.find(Action)).toHaveLength(2);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should return correct props from mapStateToProps', () => {
    const actions = [
      {
        title: 'test1',
        from: 0,
        to: 1,
        key: 0,
      },
      {
        title: 'test2',
        from: 1,
        to: 2,
        key: 1,
      },
    ];

    const initialState = {
      actions: { actions },
    };

    const result = mapStateToProps(initialState);

    expect(result).toEqual({ actions });
    expect(toJson(result)).toMatchSnapshot();
  });
});
