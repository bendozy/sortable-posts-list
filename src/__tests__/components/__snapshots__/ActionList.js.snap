// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActionList /> should render a component with action list items 1`] = `
<div
  className="ActionList"
>
  <div
    className="bg-white rounded shadow-md overflow-hidden"
  >
    <div
      className="ActionList-title font-semibold bg-white p-4 text-gray-800 text-xl"
    >
      List of actions commited
    </div>
    <div
      className="ActionList-items bg-gray-300 p-6 border border-gray-300 overflow-hidden shadow-md"
    >
      <FlipMovePropConverter
        delay={0}
        disableAllAnimations={false}
        duration={300}
        easing="ease-in-out"
        enterAnimation="elevator"
        getPosition={[Function]}
        leaveAnimation="elevator"
        maintainContainerHeight={false}
        staggerDelayBy={0}
        staggerDurationBy={0}
        typeName="div"
        verticalAlignment="top"
      >
        <Action
          action={
            Object {
              "from": 1,
              "key": 1,
              "title": "test2",
              "to": 2,
            }
          }
          index={1}
          key="1"
          timeTravel={[MockFunction]}
        />
        <Action
          action={
            Object {
              "from": 0,
              "key": 0,
              "title": "test1",
              "to": 1,
            }
          }
          index={0}
          key="0"
          timeTravel={[MockFunction]}
        />
      </FlipMovePropConverter>
    </div>
  </div>
</div>
`;

exports[`<ActionList /> should render a loading component 1`] = `
<div
  className="ActionList"
>
  <div
    className="bg-white rounded shadow-md overflow-hidden"
  >
    <div
      className="ActionList-title font-semibold bg-white p-4 text-gray-800 text-xl"
    >
      List of actions commited
    </div>
    <div
      className="ActionList-items bg-gray-300 p-6 border border-gray-300 overflow-hidden shadow-md"
    >
      <FlipMovePropConverter
        delay={0}
        disableAllAnimations={false}
        duration={300}
        easing="ease-in-out"
        enterAnimation="elevator"
        getPosition={[Function]}
        leaveAnimation="elevator"
        maintainContainerHeight={false}
        staggerDelayBy={0}
        staggerDurationBy={0}
        typeName="div"
        verticalAlignment="top"
      >
        <div
          className="ActionList-loading bg-white p-4 rounded shadow-md text-gray-500"
        >
          <span>
            No actions commited
          </span>
        </div>
      </FlipMovePropConverter>
    </div>
  </div>
</div>
`;

exports[`<ActionList /> should return correct props from mapStateToProps 1`] = `undefined`;
